build: upgrade jasmine (and related typings) to latest version

gkalpak opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] ~~Tests for the changes have been added (for bug fixes / features)~~
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


## PR Type
```
[x] Build related changes
```

## What is the current behavior?
Older version of `jasmine`/`jasmine-node` and corresponding typings.


## What is the new behavior?
Latest version of `jasmine`/`jasmine-node` and corresponding typings.


## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No(t intentionally :grin:)
```


## Other information
With these changes, the types are a little stricter now and also not compatible with Protractor's jasmine-like syntax. So, we have to also use `@types/jasminewd2` for e2e tests (but not for non-e2e tests).

I also had to "augment" `@types/jasminewd2`, because the latest typings from [DefinitelyTyped][1] do not reflect the fact that the `jasminewd2` version currently used by Protractor (v2.1.0) supports passing a `done` callback to a spec.

[1]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b7e43ef2bbfd5f045e12203f952bfe11afd03230/types/jasminewd2/index.d.ts#L9-L15

gkalpak committed:
build: upgrade jasmine (and related typings) to latest version

With these changes, the types are a little stricter now and also not
compatible with Protractor's jasmine-like syntax. So, we have to also
use `@types/jasminewd2` for e2e tests (but not for non-e2e tests).

I also had to "augment" `@types/jasminewd2`, because the latest
typings from [DefinitelyTyped][1] do not reflect the fact that the
`jasminewd2` version (v2.1.0) currently used by Protractor supports
passing a `done` callback to a spec.

[1]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/566e0394859fdc1dc893658ccec6b06372d56a91/types/jasminewd2/index.d.ts#L9-L15

Fixes #23952
Closes #24733

mary-poppins commented:
You can preview 3e2ed3c at https://pr19904-3e2ed3c.ngbuilds.io/.

gkalpak commented on tools/types-ext/jasminewd2.d.ts:
```diff
@@ -0,0 +1,35 @@
+declare global {
+  namespace jasmine {
+  // `jasmine` typings do not export `DoneFn`. Re-implement in order to use below.
+  interface DoneFn extends Function {
```
Come on clang-format! 😞

gkalpak commented on tools/types-ext/jasminewd2.d.ts:
```diff
@@ -0,0 +1,35 @@
+// Overwrite signatures to add a `done` callback.
+declare function it(
+    expectation: string, assertion?: (done: DoneFn) => Promise<void>, timeout?: number): void;
+    declare function fit(
```
😞 😞 😞

mary-poppins commented:
You can preview bc82db9 at https://pr19904-bc82db9.ngbuilds.io/.

mary-poppins commented:
You can preview c0746ed at https://pr19904-c0746ed.ngbuilds.io/.

mary-poppins commented:
You can preview c4c20ae at https://pr19904-c4c20ae.ngbuilds.io/.

mary-poppins commented:
You can preview 9316f5b at https://pr19904-9316f5b.ngbuilds.io/.

mary-poppins commented:
You can preview 5a78966 at https://pr19904-5a78966.ngbuilds.io/.

gkalpak commented on packages/common/test/pipes/date_pipe_spec.ts:
```diff
@@ -272,7 +272,7 @@ export function main() {
       // https://github.com/angular/angular/issues/9524
       it('should format correctly with iso strings that contain time',
-         () => expect(pipe.transform('2017-05-07T22:14:39', 'dd-MM-yyyy HH:mm'))
+         () => expect(pipe.transform('2017-05-07T12:13:14', 'dd-MM-yyyy HH:mm'))
```
This relies on timezone (e.g. failed on my UTC+2 timezone 😞).
Using `12` as the hour it is more resilient to timezone differences (albeit not entirely immune) and thus more likely to pass when run locally.

ocombe replied:
yes, that test is a little bit flacky

gkalpak commented on packages/compiler-cli/test/metadata/collector_spec.ts:
```diff
@@ -748,7 +749,7 @@ describe('Collector', () => {
         ]
       }]
-    });
+    } as any as MetadataMap);
```
I noticed that there are several occurrences where types were a little loose or not-specific in `compiler-cli`-related stuff. For example nullable/"undefined-able" values typed as non-nullable (e.g. [here][1], [here][2], [here][3]) or values whose properties could be more strictly typed (e.g. the changes in this file), etc.

Is this expected/intentional, @tbosch, @chuckjaz?
<sub>
Obviously, I don't understand enough about the implications of changing the types of these APIs :grin:
Just bringing it up as my instinctive, very-high-level-understanding reaction/thought.
</sub>

[1]: https://github.com/angular/angular/pull/19904/files#diff-93cb4d146755c3214602a83de1c313f2R455
[2]: https://github.com/angular/angular/pull/19904/files#diff-cdf1f6981339577e6b2003f82efb749eL118
[3]: https://github.com/angular/angular/pull/19904/files#diff-0cb86a67c64e9b1b291919678078a500R275

chuckjaz replied:
Can you leave TODO review comments anywhere where `as any as` is needed?

gkalpak commented on packages/core/test/linker/source_map_integration_node_only_spec.ts:
```diff
@@ -20,7 +20,12 @@ export function main() {
 
     beforeEach(() => {
+      // Jasmine relies on methods on `Function.prototype`, so restore the prototype on the spy.
+      // TODO: Figure out a better way to retrieve the JIT sources, without spying on `Function`.
```
@tbosch, @chuckjaz: Not a big deal, but if it is possible to retrieve the JIT sources without spying on such a fundamental built-in (`Function`), I think it is worth doing. (The current approach is a little brittle.)

gkalpak commented on packages/router/src/url_tree.ts:
```diff
@@ -111,7 +111,7 @@ export class UrlTree {
       public root: UrlSegmentGroup,
       /** The query params of the URL */
-      public queryParams: {[key: string]: string},
+      public queryParams: Params,
```
This is a public API (`{key: string}` --> `{key: any}`) plus using the already public `Params` type for that.
But I think it is necessary, because this is what actually happens (i.e. the values are not restricted to `string`).

If a more specific type is desirable, I can dig deeper into what the possible types are and update (from a quick look it might just be `string | string[]`).

IgorMinar replied:
@jasonaden will have to take a look at this. it looks like a breaking change to me, which we can't do now. Jason is this the same api change you were trying to make during v5 RC?

gkalpak replied:
Just to be clear, the public API type is wrong. It claims that `queryParams` values are always strings, but this is not the case. So, your app might typecheck fine, but break at runtime. But this change might indeed cause an app not to compile any more (for good reasons :wink:)

LMK if it is better to remove it (and use `as any` in the tests to work around it).

gkalpak replied:
Just to be more clear, `Params` is alias for `{[key: string]: any}`, so it is wider than the previous type.

gkalpak commented on tools/types-ext/jasminewd2.d.ts:
```diff
@@ -0,0 +1,35 @@
+    declare function beforeEach(action: (done: DoneFn) => Promise<void>, timeout?: number): void;
+    declare function afterEach(action: (done: DoneFn) => Promise<void>, timeout?: number): void;
+    declare function beforeAll(action: (done: DoneFn) => Promise<void>, timeout?: number): void;
+    declare function afterAll(action: (done: DoneFn) => Promise<void>, timeout?: number): void;
```
@vikerman (or anyone else working on [jasminewd][1]), should these changes be ported to [DefinitelyTyped][2]?

[1]: https://github.com/angular/jasminewd
[2]: https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jasminewd2

alexeagle replied:
Yes, you can send me a PR against the upstream typings on DefinitelyTyped, I'm a committer there

mary-poppins commented:
You can preview 8a3b7da at https://pr19904-8a3b7da.ngbuilds.io/.

mary-poppins commented:
You can preview 4e2adb2 at https://pr19904-4e2adb2.ngbuilds.io/.

mary-poppins commented:
You can preview 1e0a451 at https://pr19904-1e0a451.ngbuilds.io/.

mary-poppins commented:
You can preview b96ce54 at https://pr19904-b96ce54.ngbuilds.io/.

mary-poppins commented:
You can preview 6324f02 at https://pr19904-6324f02.ngbuilds.io/.

IgorMinar commented on packages/examples/tsconfig.json:
```diff
@@ -0,0 +1,21 @@
+// WARNING:
```
maybe call this file as tsconfig.e2e.json?

gkalpak replied:
I thought aboutt that. The only issue is that VSCode didn't seem to pick it up (not sure what other editors/IDEs do) 😒

IgorMinar commented on tools/public_api_guard/router/router.d.ts:
```diff
@@ -551,9 +551,7 @@ export declare abstract class UrlSerializer {
-    queryParams: {
-        [key: string]: string;
-    };
+    queryParams: Params;
```
this change won't fly now. we can't be changing the public api in this way.

IgorMinar reviewed (changes requested):
can we implement this without the public api change? I believe Jason is already working on a rollout plan for the router api change.

gkalpak commented:
@IgorMinar, should I still revert the public API change (despite https://github.com/angular/angular/pull/19904#discussion_r148654182)?

mary-poppins commented:
You can preview 4bba091 at https://pr19904-4bba091.ngbuilds.io/.

mary-poppins commented:
You can preview 2630492 at https://pr19904-2630492.ngbuilds.io/.

mary-poppins commented:
You can preview 593841e at https://pr19904-593841e.ngbuilds.io/.

IgorMinar commented:
I really need @jasonaden to look at the API change because he attempted to do something similar just to break g3 and revert that change. Jason?

jasonaden commented:
The change we were trying to make was to make the type `{[key: string]: string | string[]}`. I think the API change looked the same as this, but in that other change we also changed Params to the above (which is the accurate type). So this change should be okay in terms of the types because this PR is making the types more general rather than making them more specific.

gkalpak commented:
Is there a way to try this PR in g3 and see if anything breaks?

mary-poppins commented:
You can preview c6f1342 at https://pr19904-c6f1342.ngbuilds.io/.

IgorMinar commented:
@gkalpak if you get it pass reviews and CI then apply the merge label and it will be tested in g3 by the caretaker. otherwise you'd need to find a googler to test it in an ad-hoc fashion.

mary-poppins commented:
You can preview 3ff492c at https://pr19904-3ff492c.ngbuilds.io/.

mary-poppins commented:
You can preview d01cf8b at https://pr19904-d01cf8b.ngbuilds.io/.

gkalpak commented:
> @gkalpak if you get it pass reviews and CI

CI is green.
Not sure how to get reviews to pass :grin:

mary-poppins commented:
You can preview fe93a10 at https://pr19904-fe93a10.ngbuilds.io/.

mary-poppins commented:
You can preview 15f1b21 at https://pr19904-15f1b21.ngbuilds.io/.

alexeagle commented on modules/types.d.ts:
```diff
@@ -11,7 +11,9 @@
 /// <reference path="../node_modules/zone.js/dist/zone.js.d.ts" />
 /// <reference path="../node_modules/@types/hammerjs/index.d.ts" />
 /// <reference path="../node_modules/@types/jasmine/index.d.ts" />
```
at present, this means that usage of jasmine will type-check in production code. We should probably have different typings present for test compilations - but I can clean this up later as part of the bazel conversion; you're not introducing the problem

gkalpak replied:
TBH, I am not even sure if this file is used at all (or what for) 😇
Could be a relic from before moving from `modules/` to `packages/`.

alexeagle reviewed (approved):
approval for the 'bazel' group

gkalpak commented:
@alexeagle, is the approval explicitly only for the `bazel` group? What about `build-and-ci`, `compiler-cli` and `integration`? :cry:

alexeagle commented:
Click the details link for the pullapprove status. You'll see a whole
explanation of which individuals you need to chase down to get your missing
approvals. I suggest pinging them individually until you're collected the
whole set.

On Tue, Dec 12, 2017 at 6:08 AM George Kalpakas <notifications@github.com>
wrote:

> @alexeagle <https://github.com/alexeagle>, is the approval explicitly
> only for the bazel group? What about build-and-ci, compiler-cli and
> integration? 😢
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/19904#issuecomment-351060723>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC5I0QXfHSyAHWbN8Lyu9d4k5Ix4bbOks5s_oiMgaJpZM4QErCh>
> .
>

IgorMinar reviewed (approved):
I talked to Jason. all good now.

IgorMinar commented:
@gkalpak apply the merge label whenever you are ready. I'm not sure if you are still changing anything in the PR..

gkalpak commented:
I have just been rebasing the PR on master (and fixing conflicts and new errors). Applying the merge label.
Thx for the reviews!

mary-poppins commented:
You can preview 3e2ed3c at https://pr19904-3e2ed3c.ngbuilds.io/.
You can preview bc82db9 at https://pr19904-bc82db9.ngbuilds.io/.
You can preview c0746ed at https://pr19904-c0746ed.ngbuilds.io/.
You can preview c4c20ae at https://pr19904-c4c20ae.ngbuilds.io/.
You can preview 9316f5b at https://pr19904-9316f5b.ngbuilds.io/.
You can preview 5a78966 at https://pr19904-5a78966.ngbuilds.io/.
You can preview 8a3b7da at https://pr19904-8a3b7da.ngbuilds.io/.
You can preview 4e2adb2 at https://pr19904-4e2adb2.ngbuilds.io/.
You can preview 1e0a451 at https://pr19904-1e0a451.ngbuilds.io/.
You can preview b96ce54 at https://pr19904-b96ce54.ngbuilds.io/.
You can preview 6324f02 at https://pr19904-6324f02.ngbuilds.io/.
You can preview 4bba091 at https://pr19904-4bba091.ngbuilds.io/.
You can preview 2630492 at https://pr19904-2630492.ngbuilds.io/.
You can preview 593841e at https://pr19904-593841e.ngbuilds.io/.
You can preview c6f1342 at https://pr19904-c6f1342.ngbuilds.io/.
You can preview 3ff492c at https://pr19904-3ff492c.ngbuilds.io/.
You can preview d01cf8b at https://pr19904-d01cf8b.ngbuilds.io/.
You can preview fe93a10 at https://pr19904-fe93a10.ngbuilds.io/.
You can preview 15f1b21 at https://pr19904-15f1b21.ngbuilds.io/.

mary-poppins commented:
You can preview a3e36e9 at https://pr19904-a3e36e9.ngbuilds.io/.

mary-poppins commented:
You can preview cd7457a at https://pr19904-cd7457a.ngbuilds.io/.

gkalpak commented:
Rebased on master to resolve conflicts and fixed new issues in [render3/di_spec.ts][1] and [render3/properties_spec.ts][2].

[1]: https://github.com/angular/angular/pull/19904/files#diff-e2a2b6e7885a8a837c1451411c46c084
[2]: https://github.com/angular/angular/pull/19904/files#diff-91ba8623acc9f9071a8ee650ee6c4d56

gkalpak commented:
Thx for trying :smiley:

If/when someone wants to take another stab at it, lmk and I will resolve the conflicts (I won't be rebasing regularly, because it isn't much fun :grin:)
Or maybe it is not worth it and we should just close it 😕

alfaproject commented:
@gkalpak a suggestion if I may since I had to do something similar recently. I ended up creating satellite PRs of small things that I found when the types got more strict like, for example, the public API change would be a good candidate to do in another PR. The downside is that it will take longer to get this landed. #life

gkalpak commented:
Thx for the suggestion, @alfaproject. Unfortunately, I didn't do much else than upgrading jasmine (and `@types`) and fixing up the errors that came up. So, other than the public API change (which is actually a type fix, not an actual change) I don't know what else could be split out into separate PRs 😕
<sub>(And I don't think the public API change is the reason this is difficult to get merged.)</sub>

That said, I am happy to do whatever makes this more easily mergeable (from a caretaker point-of-view).

gkalpak committed:
fixup! build: upgrade jasmine (and related typings) to latest version

mary-poppins commented:
You can preview 74a80af at https://pr19904-74a80af.ngbuilds.io/.

mary-poppins commented:
You can preview d6bca63 at https://pr19904-d6bca63.ngbuilds.io/.

mary-poppins commented:
You can preview 88d620b at https://pr19904-88d620b.ngbuilds.io/.

mary-poppins commented:
You can preview 28a5abe at https://pr19904-28a5abe.ngbuilds.io/.

mary-poppins commented:
You can preview 0d81dc3 at https://pr19904-0d81dc3.ngbuilds.io/.

gkalpak committed:
build: upgrade jasmine to 2.99.x and fix tests

mary-poppins commented:
You can preview 76abb75 at https://pr19904-76abb75.ngbuilds.io/.

mary-poppins commented:
You can preview 25a5698 at https://pr19904-25a5698.ngbuilds.io/.

gkalpak commented on packages/types.d.ts:
```diff
@@ -8,20 +8,19 @@
 /// <reference path="../node_modules/@types/hammerjs/index.d.ts" />
 /// <reference path="../node_modules/@types/jasmine/index.d.ts" />
 /// <reference path="../node_modules/@types/node/index.d.ts" />
-/// <reference path="../node_modules/@types/selenium-webdriver/index.d.ts" />
```
Protractor explicitly depends on these, so we shouldn't need to manually bring them in.

gkalpak committed:
test: make `NgMatchers` type-aware

mary-poppins commented:
You can preview 8bf34b7 at https://pr19904-8bf34b7.ngbuilds.io/.

mary-poppins commented:
You can preview cbf2c48 at https://pr19904-cbf2c48.ngbuilds.io/.

gkalpak committed:
refactor: infer type for `it()` assertion functions

mary-poppins commented:
You can preview 2ae2792 at https://pr19904-2ae2792.ngbuilds.io/.

gkalpak committed:
build: upgrade jasmine to 3.1.0

mary-poppins commented:
You can preview e147306 at https://pr19904-e147306.ngbuilds.io/.

gkalpak committed:
build: upgrade karma and related dependencies

mary-poppins commented:
You can preview 1fdac95 at https://pr19904-1fdac95.ngbuilds.io/.

mary-poppins commented:
You can preview 97dfaf3 at https://pr19904-97dfaf3.ngbuilds.io/.

mary-poppins commented:
You can preview 0f35833 at https://pr19904-0f35833.ngbuilds.io/.

gkalpak committed:
refactor: re-organize and "modernize" cjs-jasmine scripts

gkalpak committed:
test: run unit tests in random order

mary-poppins commented:
You can preview 1a3f52c at https://pr19904-1a3f52c.ngbuilds.io/.

mary-poppins commented:
You can preview b3a3be0 at https://pr19904-b3a3be0.ngbuilds.io/.

mary-poppins commented:
You can preview 8071de3 at https://pr19904-8071de3.ngbuilds.io/.

mary-poppins commented:
You can preview 0ebba85 at https://pr19904-0ebba85.ngbuilds.io/.

gkalpak commented:
This is now ready for review 😅
(cc @mhevery, @rkirov)

mhevery commented:
http://test/OCL:203491123:BASE:203491196:1530896473331:c6268e84

alfaproject commented on packages/benchpress/test/webdriver/chrome_driver_extension_spec.ts:
```diff
@@ -6,7 +6,7 @@
  */
 
-import {AsyncTestCompleter, describe, expect, iit, inject, it} from '@angular/core/testing/src/testing_internal';
+import {AsyncTestCompleter, describe, expect, fit, inject, it} from '@angular/core/testing/src/testing_internal';
```
Oops?

mhevery replied:
fixed

gkalpak replied:
What was the problem??? 😕

gkalpak replied:
AFAICT, everything was fine before (and Travis agrees with me 😛).

mhevery replied:
Yes, had to revert my `fit` => `iit` change.

alfaproject commented on packages/common/http/test/client_spec.ts:
```diff
@@ -6,7 +6,7 @@
  */
 
-import {ddescribe, describe, iit, it} from '@angular/core/testing/src/testing_internal';
+import {ddescribe, describe, fit, it} from '@angular/core/testing/src/testing_internal';
```
Oops?

mhevery replied:
fixed

gkalpak replied:
😕

mary-poppins commented:
You can preview edb5217 at https://pr19904-edb5217.ngbuilds.io/.
